README.md
# ![wellid](http://broeser.github.io/wellid/images/wellid-logo.svg)

wellid is a set of PHP validators and a few loosely coupled components for validation.

[![Build Status](https://travis-ci.org/broeser/wellid.svg?branch=master)](https://travis-ci.org/broeser/wellid)
[![codecov.io](https://codecov.io/github/broeser/wellid/coverage.svg?branch=master)](https://codecov.io/github/broeser/wellid?branch=master)
[![License](http://img.shields.io/:license-mit-blue.svg)](http://mit-license.org)
[![SemVer 2.0.0](https://img.shields.io/badge/semver-2.0.0-blue.svg)](http://semver.org/spec/v2.0.0.html)

Latest stable version: 0.4.0

## Goals

- wellid should be easy to use and easy to learn
- It is up to you, how much of the wellid-package you use, you can start with
  small building blocks and use bigger concepts later
- Extending wellid with your own Validators should be easy 

## Installation

wellid works with PHP 5.6 and 7.0.

The package can be installed via composer:

``composer require broeser/wellid``

## Before you start

- All examples from this manual can be found in [usage_examples.php](usage_examples.php)
  in the same order as in the manual. If an example uses an additional class,
  that class can be found in the examples/-directory.
- **IMPORTANT NOTE:** Never try to validate raw data! Sanitize your data first, 
then pass it to wellid. Recommended sanitization options are:
  1. Let your framework handle sanitization
  2. Use [Sanitor](https://github.com/broeser/sanitor) (composer require broeser/sanitor)
  3. Use PHP's filter_input() and filter_var() methods

## Using wellid

### Simple use case _with validateBool()_

```PHP
<?php
/*
 * The value to validate
 */
$value = 'somethingentered';

/*
 * The validator that shall be used to validate it
 */
$maxLengthValidator = new Wellid\Validator\MaxLength(7);

/*
 * Validate the value and do something in case it is valid:
 */
if($maxLengthValidator->validateBool($value)) {
    print('The given value '.$value.' fits our requirements of a maximum length of 7 characters! YAY!'.PHP_EOL);
} else {
    print('The given value '.$value.' is totally invalid. :-('.PHP_EOL);
}
```

The simplest usage case is creating a new validator and using the 
**validateBool()**-method. It takes the value that shall be validated as parameter
and returns true on success and false on failure.

These are the validators supplied with wellid by default:

 - Boolean
 - Date
 - Email
 - Filesize
 - FloatingPoint
 - IPAddress
 - Integer
 - MacAddress
 - Max
 - MaxDate
 - MaxLength
 - MIME
 - Min
 - MinDate
 - MinLength
 - Password
 - URL

### Error handling _with validate() and ValidationResult_

Sometimes it is important to know, _why_ validation failed. If you need more than
boolean true/false, you can use the **validate()**-method of a validator of
your choice to get a **ValidationResult**-object. A ValidationResult includes 
an error message and error code if validation fails, you can use **getCode()** and
 **getMessage()** to retrieve them:

```PHP
<?php
/*
 * The value to validate
 */
$value = 'somethingentered';

/*
 * The validator that shall be used to validate it
 */
$maxLengthValidator = new Wellid\Validator\MaxLength(7);

/*
 * Validate the value and get a ValidationResult-object
 */
$validationResult = $maxLengthValidator->validate($value);

if($validationResult->hasPassed()) {
    print('The given value '.$value.' fits our requirements of a maximum length of 7 characters! YAY!'.PHP_EOL);
} else {
    print('The given value '.$value.' is totally invalid. :-('.PHP_EOL);
    print('Reason: '.$validationResult->getMessage().' – Error code: '.$validationResult->getCode().PHP_EOL);
}
```

### ValidationResultSets _– A collection of ValidationResults_

If you validate a value with several different validators, or if you validate a
lot of different values you might need a better way of handling the results:

You can combine several ValidationResults to form a ValidationResultSet. You can
**add()** a single ValidationResult to a ValidationResultSet or you can combine
two ValidationResultSets with **addSet()**

ValidationResultSets can be count()-ed and iterated over with foreach(). 

**hasErrors()** is useful to check if there are any errors
in the ValidationResultSet, its counterpart is called **hasPassed()**. You can 
retrieve the **firstError()**-ValidationResult (if there is any), if everything
has passed the method will return null.

```PHP
<?php
/*
 * The value to validate
 */
$value = 'somethingentered';

/*
 * The validator that shall be used to validate it
 */
$maxLengthValidator = new Wellid\Validator\MaxLength(7);
$minLengthValidator = new Wellid\Validator\MinLength(3);

$validationResultSet = new Wellid\ValidationResultSet();
$validationResultSet->add($maxLengthValidator->validate($value));
$validationResultSet->add($minLengthValidator->validate($value));

if($validationResultSet->hasPassed()) {
    print('The given value '.$value.' fits our requirements of a maximum and minimum length! YAY!'.PHP_EOL);
} else {
    print('The given value '.$value.' is totally invalid. :-('.PHP_EOL);
}
```


### Validating objects _with ValidatableTrait and ValidatableInterface_

If you want to create classes (as opposed to primitive data types) whose instances
are validateable by wellid, just implement the **ValidatableInterface** in your class
and use the **ValidatableTrait**. If you prefer abstract classes instead, (e.g.
if you want to override functionality from ValidatableTrait), extending
**AbstractValidatable** is the way to go. In either case, Make sure to implement
 a **getValue()** method to supply the validators with a primitive typed version
of your object's value. For a Money-class, for example, that might be a float.

You'll now be able to **addValidators()** to your 
object. You can validate your object with the **validate()** and **validateBool()**
methods. 

Note that **validate()** returns a ValidationResultSet and not a ValidationResult
(see above).

The following example code uses three validators: 
The value shall be a floating point number. It shall be between 0 (zero) and 
830.

```PHP
<?php
class AccountBalance implements \Wellid\ValidatableInterface {
    use \Wellid\ValidatableTrait;

    /**
     * @var float
     */
    protected $value = null;

    public function __construct() {
        $this->addValidators(new \Wellid\Validator\FloatingPoint(), new \Wellid\Validator\Min(0), new \Wellid\Validator\Max(830));
    }
    
    /**
     * @param float $val
     * @return \self
     */    
    public static function createFromFloat($val) {
        $newInstance = new self();
        $newInstance->setValue($val);
        return $newInstance;
    }
    
    /**
     * @return float
     */
    public function getValue() {
        return $this->value;
    }
    
    /**
     * @param float $val
     */
    public function setValue($val) {
        $this->value = $val;
    }
}

foreach(array(57.3, -6) as $v) {
    $yourBalance = WellidUsageExamples\AccountBalance::createFromFloat($v);
    $result = $yourBalance->validate();
    if($result->hasErrors()) {
        print('Oh dear! Something invalid was used as my account balance!'.PHP_EOL);
        print('Aha, that is why: '.$result->firstError()->getMessage().PHP_EOL);
    }
}
```

### Caching ValidationResultSets _with CacheableValidatableTrait and CacheableValidatableInterface_

If a lot of Validators are used in validating an object, caching might improve
performance. In the last chapter each call to validate() or validateBool()
starts validation anew. To add caching functionality use 
**CacheableValidatableInterface** instead of ValidatableInterface and 
**CacheableValidatableTrait** instead of ValidatableTrait. If you prefer an
abstract class over traits and interfaces, extend **AbstractCacheableValidatable**.
All cache-related functionality can be found in the \Wellid\Cache-namespace.

Like the cacheless variant, CacheableValidatable has to implement the 
getValue()-method.

Caching will be performed automatically.

If you want to disable caching for a particular instance of 
CacheableValidatableInterface, you can call **disableValidationCache()**. This
will also remove a potentially existing ValidationResultSet from the cache of
that instance.

While rarely useful, you can also force revalidation without disabling the cache.
Clear the current ValidationResultSet with **clearValidationResult()**, then use
validate() or validateBool() to get a new ValidationResultSet.


### A collection of Validators _– The ValidatorHolder_

The way wellid is designed, you can always add validators directly to your data
objects. However it might become handy to store a collection of validators
separate from the data objects or even without having data objects.

You can use the **ValidatorHolder**-class directly, extend it by your own class,
or you can create a class that implements **ValidatorHolderInterface** and may
use the **ValidatorHolderTrait** to get some basic functionality.

Adding validators works the same as on data objects: You
can use addValidators() or **addValidator()** (for a single validator). Use 
**getValidators()** to retrieve an array of all assigned validators.

To validate a value with the ValidatorHolder, use the **validateValue()**-method. 

The AccountBalance-example from above becomes much cleaner and easier to 
understand. 

Example:

```PHP
<?php
class AccountBalanceValidators extends \Wellid\ValidatorHolder {
    public function __construct() {
        $this->addValidators(new \Wellid\Validator\FloatingPoint(), new \Wellid\Validator\Min(0), new \Wellid\Validator\Max(830));
    }
}

$accountBalanceValidators = new \WellidUsageExamples\AccountBalanceValidators();
foreach(array(57.3, -6) as $v) {
    $result = $accountBalanceValidators->validateValue($v);
    if($result->hasErrors()) {
        print('Oh dear! Something invalid was used as my account balance!'.PHP_EOL);
        print('Aha, that is why: '.$result->firstError()->getMessage().PHP_EOL);
    }
}
```
Of course you don't have to setup the validators in the constructor in your own 
project, but you can just call addValidator or addValidators from anywhere.

Please note, that there is currently no validateBoolValue()-method for
ValidatorHolders. If you need the boolean value you can use the syntax
``$validationResultAsBool = $accountBalanceValidators->validateValue($value)->hasPassed();``

Another example shows, how to use ValidatorHolders with data objects. (Both 
used classes are the same as in the examples above):
```PHP
<?php
/*
 * Example 3b: Using the ValidatorHolderTrait & ValidatorHolderInterface with
 * data objects
 */
$accountBalanceValidators = new \WellidUsageExamples\AccountBalanceValidators();
foreach(array(57.3, -6) as $v) {
    $yourBalance = WellidUsageExamples\AccountBalance::createFromFloat($v);
    $result = $accountBalanceValidators->validateValue($yourBalance->getValue());
    if($result->hasErrors()) {
        print('Oh dear! Something invalid was used as my account balance!'.PHP_EOL);
        print('Aha, that is why: '.$result->firstError()->getMessage().PHP_EOL);
    }
}
```

### Using wellid with Sanitor _with SanitorBridgeTrait and SanitorBridgeInterface_

[Sanitor](https://github.com/broeser/sanitor) is a thin wrapper around PHP's 
sanitization functions (filter_var(), filter_input(), etc.). 
If you want to use Sanitor to sanitize input or arbitrary
values before validating them with wellid, there is a handy piece of code called
the **SanitorBridgeTrait** just for that.

Refer to Sanitor's README.md for more information.
These are the four basic steps necessary to integrate Sanitor and wellid

1. Install the Sanitor package (composer require broeser/sanitor)
2. Substitute the ValidatableTrait and CacheableValidatableTraits with the 
   **SanitorBridgeTrait** in all places.
3. Make sure that those classes implement **SanitorBridgeInterface**
4. Make sure that those classes call $this->setSanitizer(...) somewhere
   before validation (e. g. in the constructor) and set a fitting
   sanitization filter (you can try FILTER_DEFAULT)

SanitorBridge automatically uses caching. You don't have to clear the cache
when setting a new rawValue, this will be done automatically.

```PHP
<?php
class SanitorWellidEmailExample implements \Wellid\SanitorBridgeInterface, \Wellid\ValidatableInterface {
    use \Wellid\SanitorBridgeTrait, \Wellid\ValidatableTrait;

    /**
     * Constructor
     * 
     * @param \Sanitor\Sanitizer $sanitizer
     */
    public function __construct(\Sanitor\Sanitizer $sanitizer = null) {
        $this->setSanitizer(is_null($sanitizer)?new \Sanitor\Sanitizer(FILTER_SANITIZE_EMAIL):$sanitizer);
        $this->addValidator(new \Wellid\Validator\Email());
    }
}

$emailValidator = new WellidUsageExamples\SanitorWellidEmailExample(new \Sanitor\Sanitizer(FILTER_SANITIZE_EMAIL));
$emailValidator->setRawValue('mail@benedict\roeser.de'); // because the value will be sanitized before validation, this will actually pass! More information in the "65{"-example below
if($emailValidator->validate()->hasErrors()) {
    print('Why! Oh why! Errors everywhere!'.PHP_EOL);
}

// values can also be optained from INPUT_GET, INPUT_POST, etc.:
$emailValidator->rawValueFromInput(INPUT_REQUEST, 'email');
if($emailValidator->validate()->hasPassed()) {
    print('Nice, this input has been valid: '.$emailValidator->getValue().PHP_EOL);
}
```

Imagine you are expecting an integer as value, but the user enters `65{` instead.
Depending on your business logic, two different cases are possible:
 
 1. Ignore the { and assume 65, continue working with 65. Optionally notify the
    user of this – this makes sense if there is an undo anyway and you don't
    want to annoy your users with error messages.
 2. Return an error message and ask for the value again – this makes sense if
    the user is not expected to notice/fix the mistake, or if valid data is
    more important than user experience

The first case is the default setting.

For the second case, just add a **SanitorMatch**-Validator to your object before
starting validation. The SanitorMatch-Validator expects the validatable object 
itself as parameter on construction. The Validator uses 
"The given value contains illegal characters" as error message, if validation
fails.

```PHP
<?php
$emailValidator->addValidator(new \Wellid\Validator\SanitorMatch($emailValidator));
```

Luckily there is a shorter way to accomplish the same with the method
**addSanitorMatchValidator()**:
```PHP
<?php
$emailValidator->addSanitorMatchValidator();
```

### Exceptions

Feel free to use the Exception-classes supplied with wellid in any validation
context you want to.

 - DataFormat
 - DataType
 - NotFound
 - FileNotFound

## Contributing?

Yes, please!

See [CONTRIBUTING.md](CONTRIBUTING.md) for details and/or open an issue with your questions.

Please note that this project is released with a [Contributor Code of Conduct](CODE_OF_CONDUCT.md). 
By participating in this project you agree to abide by its terms.

## wellid?

Yes, it is a pun on well and valid.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.
This project adheres to [Semantic Versioning](http://semver.org/).

## [Unreleased]

## [0.4.0] - 2016-02-14
### Added
- **MinDate** and **MaxDate**-Validators to validate if a date is within a given
  range

### Fixed
- **Integer** and **Float**-Validators did not work correctly for 0

## [0.3.0] - 2016-02-03
### Added
- **FloatingPoint**-Validator now supports optional decimal separator parameter
  in constructor
- **CacheableValidatable…** (in Cache/) that allow to cache ValidationResultSets;
  available as combination of trait and interface or as abstract class
- **AbstractValidatable** can be used as an alternative to ValidatableTrait and
  ValidatableInterface; Extend AbstractValidatable if you prefer an abstract base
  class, which also allows to override methods of the ValidatableTrait
- **getErrorMessages()** retrieves an array of error message strings from a
  ValidationResultSet
- **SanitorMatch**-Validator, checks whether given input stays the same, when it
  is sanitized
- **FileNotFound**-Exception that extends the NotFound-exception
- Builds are now tested automatically by Travis CI
- Added CHANGELOG, CONTRIBUTING document and CODE_OF_CONDUCT

### Changed
- **MIME**-Validator now takes a filename string instead of an array containing
  this string
- **Filesize**-Validator now takes a filename string instead of an array containing
  the size of the file
- **Date**-Validator now throws an exception, if neither NULL nor a string is 
  passed as data format
- Improved exceptions thrown by the **Password**-Validator

### Fixed
- **Boolean**-Validator now works correctly on objects and NULL
- Renamed former Float-Validator to **FloatingPoint**, in order to work on PHP 7
- **rawValueFromInput()** did not work correctly with INPUT_ENV
- Passing a non-boolean as first parameter to **ValidationResult** now throws
  the correct exception

## [0.2.0] - 2016-01-23
### Added
 - **Boolean**-Validator
 - **IPAddress**-Validator
 - **MacAddress**-Validator
 - **validateBool()**-method for Validatable…, works like validateBool() on
   Validators

### Fixed
 - **Integer**-Validator is working again

### Removed
 - **Required**-Validator

## [0.1.0] - 2016-01-22
### Added
- **firstError()**-method to get the first error in a ValidationResultSet
- **validateValue($v)**-method in Validatable… to validate specific values
- Adapter **SanitorBridge…**: Optional possibility to use wellid in conjunction 
  with Sanitor


## [0.0.2] - 2016-01-22
### Fixed
- **Date**-Validator was not working due to namespaceing issues

## [0.0.1] - 2016-01-21
- Initial release

[Unreleased]: https://github.com/broeser/wellid/compare/0.4.0...HEAD
[0.4.0]: https://github.com/broeser/wellid/releases/tag/0.4.0
[0.3.0]: https://github.com/broeser/wellid/releases/tag/0.3.0
[0.2.0]: https://github.com/broeser/wellid/releases/tag/0.2.0
[0.1.0]: https://github.com/broeser/wellid/releases/tag/0.1.0
[0.0.2]: https://github.com/broeser/wellid/releases/tag/0.0.2
[0.0.1]: https://github.com/broeser/wellid/releases/tag/0.0.1

usage_examples.php
<?php

/*
 * These are all examples that are used within the README.md. This is, however
 * NOT the README itself. It is recommended to read the README first or in
 * parallel while looking at this code.
 */

// Autoloader, provided by composer
// Configure in composer.json
require_once __DIR__.'/vendor/autoload.php';

/**
 * SECTION 1
 * Simple use case _with validateBool()_
 */

/*
 * The value to validate
 */
$value = 'somethingentered';

/*
 * The validator that shall be used to validate it
 */
$maxLengthValidator = new Wellid\Validator\MaxLength(7);

/*
 * Validate the value and do something in case it is valid:
 */
if($maxLengthValidator->validateBool($value)) {
    print('The given value '.$value.' fits our requirements of a maximum length of 7 characters! YAY!'.PHP_EOL);
} else {
    print('The given value '.$value.' is totally invalid. :-('.PHP_EOL);
}

/*
 * The simplest usage case is creating a new validator and using the 
 * __validateBool()__-method. It takes the value that shall be validated as parameter
 * and returns true on success and false on failure.
 */

/**
 * SECTION 2
 * Error handling _with validate() and ValidationResult_
 */

/*
 * Sometimes it is important to know, _why_ validation failed. If you need more than
 * boolean true/false, you can use the __validate()__-method of a validator of
 * your choice to get a __ValidationResult__-object. A ValidationResult includes 
 * an error message and error code if validation fails, you can use __getCode()__ and
 *  __getMessage()__ to retrieve them:
 */

/*
 * The value to validate
 */
$value = 'somethingentered';

/*
 * The validator that shall be used to validate it
 */
$maxLengthValidator = new Wellid\Validator\MaxLength(7);

/*
 * Validate the value and get a ValidationResult-object
 */
$validationResult = $maxLengthValidator->validate($value);

if($validationResult->hasPassed()) {
    print('The given value '.$value.' fits our requirements of a maximum length of 7 characters! YAY!'.PHP_EOL);
} else {
    print('The given value '.$value.' is totally invalid. :-('.PHP_EOL);
    print('Reason: '.$validationResult->getMessage().' – Error code: '.$validationResult->getCode().PHP_EOL);
}

/**
 * SECTION 3
 * 
 * ValidationResultSets _– A collection of ValidationResults_
 */

/**
 * If you validate a value with several different validators, or if you validate a
 * lot of different values you might need a better way of handling the results:
 * You can combine several ValidationResults to form a ValidationResultSet. You can
 * __add()__ a single ValidationResult to a ValidationResultSet or you can combine
 * two ValidationResultSets with __addSet()__
 * ValidationResultSets can be count()-ed and iterated over with foreach(). 
 * __hasErrors()__ is useful to check if there are any errors
 * in the ValidationResultSet, its counterpart is called __hasPassed()__. You can 
 * retrieve the __firstError()__-ValidationResult (if there is any), if everything
 * has passed the method will return null.
 */

/*
 * The value to validate
 */
$value = 'somethingentered';

/*
 * The validator that shall be used to validate it
 */
$maxLengthValidator = new Wellid\Validator\MaxLength(7);
$minLengthValidator = new Wellid\Validator\MinLength(3);

$validationResultSet = new Wellid\ValidationResultSet();
$validationResultSet->add($maxLengthValidator->validate($value));
$validationResultSet->add($minLengthValidator->validate($value));

if($validationResultSet->hasPassed()) {
    print('The given value '.$value.' fits our requirements of a maximum and minimum length! YAY!'.PHP_EOL);
} else {
    print('The given value '.$value.' is totally invalid. :-('.PHP_EOL);
}

/**
 * SECTION 4
 * 
 * Validating objects _with ValidatableTrait and ValidatableInterface_
 */

/**
 * If you want to create classes (as opposed to primitive data types) whose instances
 * are validateable by wellid, just implement the __ValidatableInterface__ in your class
 * and use the __ValidatableTrait__. You'll now be able to __addValidators()__ to your 
 * object. You can validate your object with the __validate()__ and __validateBool()__
 * methods. Make sure to implement a __getValue()__ method to supply the validators
 * with a primitive version of your object's value.
 * Note that __validate()__ returns a ValidationResultSet and not a ValidationResult
 * (see above).
 * 
 * The AccountBalance-class can be found in examples/AccountBalance.php. 
 * 
 * The example uses three validators: The value shall be a floating point number. 
 * It shall be between 0 (zero) and 830.
 */

foreach(array(57.3, -6) as $v) {
    $yourBalance = WellidUsageExamples\AccountBalance::createFromFloat($v);
    $result = $yourBalance->validate();
    if($result->hasErrors()) {
        print('Oh dear! Something invalid was used as my account balance!'.PHP_EOL);
        print('Aha, that is why: '.$result->firstError()->getMessage().PHP_EOL);
    }
}

/**
 * SECTION 6
 * 
 * A collection of Validators _The ValidatorHolder_*/

/*
 * The way wellid is designed, you can always add validators directly to your data
 * objects. However it might become handy to store a collection of validators
 * separate from the data objects or even without having data objects.
 * You can use the __ValidatorHolder__-class directly, extend it by your own class,
 * or you can create a class that implements __ValidatorHolderInterface__ and may
 * use the __ValidatorHolderTrait__ to get some basic functionality.
 * Adding validators works the same as on data objects: You
 * can use addValidators() or __addValidator()__ (for a single validator). Use 
 * __getValidators()__ to retrieve an array of all assigned validators.
 * To validate a value with the ValidatorHolder, use the __validateValue()__-method. 
 * The AccountBalance-example from above becomes much cleaner and easier to 
 * understand. The following example code can be found in 
 * examples/AccountBalanceValidators.php. (Of course you don't have to setup the
 * validators in the constructor in your own project, but you can just call
 * addValidator or addValidators from anywhere.)
 */

$accountBalanceValidators = new \WellidUsageExamples\AccountBalanceValidators();
foreach(array(57.3, -6) as $v) {
    $result = $accountBalanceValidators->validateValue($v);
    if($result->hasErrors()) {
        print('Oh dear! Something invalid was used as my account balance!'.PHP_EOL);
        print('Aha, that is why: '.$result->firstError()->getMessage().PHP_EOL);
    }
}

/**
 * Please note, that there is currently no validateBoolValue()-method for
 * ValidatorHolders. If you need the boolean value you can use the syntax
 *    validateValue($value)->hasPassed()
 * Another example shows, how to use ValidatorHolders with data objects. (Both 
 * used classes are the same as in the examples above):
 */

$accountBalanceValidators = new \WellidUsageExamples\AccountBalanceValidators();
foreach(array(57.3, -6) as $v) {
    $yourBalance = WellidUsageExamples\AccountBalance::createFromFloat($v);
    $result = $accountBalanceValidators->validateValue($yourBalance->getValue());
    if($result->hasErrors()) {
        print('Oh dear! Something invalid was used as my account balance!'.PHP_EOL);
        print('Aha, that is why: '.$result->firstError()->getMessage().PHP_EOL);
    }
}

/**
 * SECTION 7
 * 
 * Using wellid with Sanitor _– the SanitorBridgeTrait_
 */

/*
 * Sanitor is a thin wrapper around PHP's sanitization functions (filter_var(), 
 * filter_input(), etc.). If you want to use Sanitor to sanitize input or arbitrary
 * values before validating them with wellid, there is a handy piece of code called
 * the __SanitorBridgeTrait__ just for that.
 * Refer to the README.md in the broeser/sanitor-package for more information.
 * These are the four basic steps necessary to integrate Sanitor and wellid
 * 1. Install the Sanitor package (composer require broeser/sanitor)
 * 2. Substitute the ValidatableTrait and CacheableValidatableTraits with the 
 *    __SanitorBridgeTrait__ in all places.
 * 3. Make sure that those classes implement __SanitorBridgeInterface__
 * 4. Make sure that those classes call $this->setSanitizer(...) somewhere
 *    before validation (e. g. in the constructor) and set a fitting
 *    sanitization filter (you can try FILTER_DEFAULT)
 * The following example code can be found in examples/SanitorWellidEmailExample.php. 
 */

$emailValidator = new WellidUsageExamples\SanitorWellidEmailExample(new \Sanitor\Sanitizer(FILTER_SANITIZE_EMAIL));
$emailValidator->setRawValue('mail@benedict\roeser.de'); // because the value will be sanitized before validation, this will actually pass! More information in the "65{"-example below
if($emailValidator->validate()->hasErrors()) {
    print('Why! Oh why! Errors everywhere!'.PHP_EOL);
}

// values can also be optained from INPUT_GET, INPUT_POST, etc.:
$emailValidator->rawValueFromInput(INPUT_REQUEST, 'email');
if($emailValidator->validate()->hasPassed()) {
    print('Nice, this input has been valid: '.$emailValidator->getValue().PHP_EOL);
}

/**
 * Imagine you are expecting an integer as value, but the user enters `65{` instead.
 * Depending on your business logic, two different cases are possible:
 *  
 *  1. Ignore the { and assume 65, continue working with 65. Optionally notify the
 *     user of this – this makes sense if there is an undo anyway and you don't
 *     want to annoy your users with error messages.
 *  2. Return an error message and ask for the value again – this makes sense if
 *     the user is not expected to notice/fix the mistake, or if valid data is
 *     more important than user experience
 * The first case is already possible with the example above. For the second case,
 * just add a __SanitorMatch__-Validator to your object before starting validation.
 * The SanitorMatch-Validator expects the validatable object itself. It uses
 * "The given value contains illegal characters" as error message, if validation
 * fails.
 */
$emailValidator->addValidator(new \Wellid\Validator\SanitorMatch($emailValidator));

/**
 * Luckily there is a shorter way to accomplish the same with 
 * __addSanitorMatchValidator()__:
 */
$emailValidator->addSanitorMatchValidator();